feature: extension preview command #1505

gornostal opened the pull request:
See #1505

friday commented on ulauncher/modes/extensions/extension_cli_handlers.py:
```diff
@@ -130,3 +133,143 @@ def upgrade_extensions(_: ArgumentParser, args: Namespace) -> bool:
+    """
+    # Check if Ulauncher is running
+    if not check_app_running(app_id):
+        logger.error("Error: Ulauncher is not running. Please start it first.")
```
Nit: I don't like "Please start it first". Maybe "Error: Ulauncher needs to be running in order to preview extensions." or something.

gornostal replied:
Fixed

friday commented on ulauncher/modes/extensions/extension_mode.py:
```diff
@@ -155,3 +155,104 @@ def trigger_event(self, event: dict[str, Any]) -> None:
+        # Try to restart the original extension
+        try:
+            original_controller = extension_registry.get(original_ext_id)
+        except ExtensionNotFoundError:
```
With the changes made to the get method, this should be `except KeyError` right?

friday reviewed (changes requested):
Overall great work, but I found some issues while testing and made some code comments. I'm not sure exactly how the debugging is supposed to work because I never debug python. I didn't get it to work, but it's likely I'm missing something.

I tested with for example https://github.com/Ulauncher/ulauncher-hash which I have installed already and have the source code locally. While it was running and enabled I ran the preview command which disabled it and started in debug mode instead (as expected). But:

1. If you save the preferences then restart Ulauncher, then the app wouldn't start next time (fixed in main here 66ada59e, but not in this branch yet until you rebase it).

<img width="2486" height="884" alt="Image" src="https://github.com/user-attachments/assets/032a874f-3ea5-4014-b103-074743d8899e" />

This was really caused by a bad assumption in main from the previous changes though.

2. I managed to enabling the disabled extension manually in the preferences and for example got both of them running at the same time:

<img width="1360" height="360" alt="Image" src="https://github.com/user-attachments/assets/34b2764b-8e25-4a5d-9637-c7a1f52cb5b8" />

I don't think this should be possible. The "on" button for the overridden extension should be disabled imo, and "reload extensions" button should ignore the overridden extension.

3. If you run with the `--with-debugger`, press ctrl+c and then run with `--with-deubgger` again it errors with this message: `RuntimeError: Can't listen for client connections: [Errno 98] Address already in use  🔌 from com.github.ulauncher.ulauncher-hash.preview`

4. Also seems to only affect with debug:
I get this message and the preview extension view shows as "stopped" in the list view but not the extension details
<img width="2456" height="296" alt="Image" src="https://github.com/user-attachments/assets/1d23963b-d6b2-4608-9fb6-e860a40630da" /> 

<img width="1996" height="1076" alt="Image" src="https://github.com/user-attachments/assets/cad7bece-f7cd-4754-848c-b7fb05f95e12" />

5. If Ulauncher is terminated, the preview process stays open, but then if you start Ulauncher again, it won't "reconnect" to the preview process, so in this case you have to ctrl+c and then restart. I was running both Ulauncher and the preview command from the same terminal window but different panes, so I ended up doing this accidentally, but still seems like a legit case. Not fully sure how to support it though. Maybe get the ulauncher pid when the preview runtime start and check it regularly and kill the preview process if the ulauncher pid dies.

gornostal commented:
Thanks for the review!
Very thorough. As always :) 

1. Rebased
2. Fixed. Now you won't be able to enable it or restart by clicking the reload button
3. Fixed. It was a bug in `run_ext_batch_job`
4. That's expected because the extension is waiting for a debug client to connect before executing the first line of code.
   I added a note to the pref. UI so the user knows what to do next.
5. Fixed. Same as point 3.

gornostal commented on ulauncher/modes/extensions/extension_mode.py:
```diff
@@ -65,7 +65,7 @@ def handle_action(self, action_metadata: ActionMetadata | None) -> None:
         self._trigger_cache.clear()
         for ext in extension_registry.iterate():
-            if not ext.is_enabled or ext.has_error:
+            if not ext.is_enabled or ext.has_error or not ext.is_running:
```
Pushed to `main`

friday reviewed (commented):
I tried to make time yesterday to review, but I didn't have time to test. The code looks good now. One concern is that I think you didn't disable or hide the "on" button for the `shadowed_by_preview` I think? So there is no indication that it's disabled. I'll try to test it tomorrow,

friday commented:
I just re-read and might add one more change request (small one).

gornostal commented:
Thanks for taking the time for this!
